### PR TITLE
vim-patch:8.2.4014: git and gitcommit file types not properly recognized

### PR DESCRIPTION
**Pull request by clason**
Problem:    Git and gitcommit file types not properly recognized.
Solution:   Adjust filetype detection. (Tim Pope, closes vim/vim#9477)
https://github.com/vim/vim/commit/c689f8c3d98fffe7e13730e198ce120934528f9c

### CODE REVIEW

**Comment by zeertzjq** [comment]
Hmm, #16833 appeared again?

File: src/nvim/testdir/test_filetype.vim [hunk 1]
```diff
@@ -195,7 +195,7 @@ let s:filename_checks = {
     \ 'gedcom': ['file.ged', 'lltxxxxx.txt', '/tmp/lltmp', '/tmp/lltmp-file', 'any/tmp/lltmp', 'any/tmp/lltmp-file'],
     \ 'gemtext': ['file.gmi', 'file.gemini'],
     \ 'gift': ['file.gift'],
-    \ 'gitcommit': ['COMMIT_EDITMSG', 'MERGE_MSG', 'TAG_EDITMSG'],
+    \ 'gitcommit': ['COMMIT_EDITMSG', 'MERGE_MSG', 'TAG_EDITMSG', 'NOTES_EDITMSG', 'EDIT_DESCRIPTION'],
```
Comment (seandewar) [inline src/nvim/testdir/test_filetype.vim:198]: This'll need to be added to `filetype_spec.lua`

Comment (clason) [reply src/nvim/testdir/test_filetype.vim:198]: Right, forgot we added a new test!

**Review by seandewar — APPROVED** [review]

**Comment by seandewar** [comment]
Let's see if we trigger #16833 again :eyes:

File: runtime/filetype.vim [hunk 2]
```diff
@@ -680,12 +681,8 @@ if !empty($XDG_CONFIG_HOME)
...
 au BufRead,BufNewFile .gitsendemail.msg.??????	setf gitsendemail
-au BufNewFile,BufRead .msg.[0-9]*
-      \ if getline(1) =~ '^From.*# This line is ignored.$' |
-      \   setf gitsendemail |
-      \ endif
 au BufNewFile,BufRead *.git/*
-      \ if getline(1) =~ '^\x\{40\}\>\|^ref: ' |
+      \ if getline(1) =~# '^\x\{40,\}\>\|^ref: ' |
       \   setf git |
       \ endif
```
Comment (gpanders) [inline runtime/filetype.vim:687]: Do you want to include a port of this into filetype.lua?

I think this ought to do it?

```lua
[".*%.git/.*"] = function(path, bufnr)
  local firstline = getline(1)
  if firstline:find("^" .. string.rep("%x", 40) .. "+ ") or firstline:sub(1, 5) == "ref: " then
    return "git"
  end
end,
```

**Review by gpanders — APPROVED** [review]

File: runtime/filetype.vim (hunk 2 above)
Comment (clason) [reply runtime/filetype.vim:687]: Worth a try -- I trust you on the Lua regex! What about the test?

Comment (clason) [reply runtime/filetype.vim:687]: Or maybe better to include it in https://github.com/neovim/neovim/pull/16959#pullrequestreview-845786399 (which also needs a similar test).

**Comment by zeertzjq** [comment]
I tried applying the patch on `master` and LSAN surprisingly didn't fail, so let's try a rebase.

**Comment by zeertzjq** [comment]
Finally ASAN passed.

File: runtime/filetype.vim (hunk 2 above)
Comment (clason) [reply runtime/filetype.vim:687]: I'm merging this now while ASAN is still green, so the `git` pattern should go into the linked PR.

**Pull request merged**